### PR TITLE
Remote Messaging

### PR DESCRIPTION
**Pull request by amddg44**
Task/Issue URL: https://app.asana.com/0/72649045549333/1201884213179453/f
Tech Design URL: https://app.asana.com/0/481882893211075/1202238116542877/f
CC:

**Description**:
This PR is for the implementation of Remote Messaging.

**Steps to test this PR**:
1. No message during onboarding
	• Fresh install
	• Open the app and go through the onboarding screens
	• Ensure you don't see any remote message on the home tab if there's a Dax dialog
	• after completing the onboarding, relaunch the app and ensure first remote message appears
	• ensure pixel `m_remote_message_shown` and `m_remote_message_shown_unique`
	• visit any site, and open a new tab
	• Ensure message appears on the new home tab

2. New message
	• Now click on the "X" to dismiss the remote message
	• ensure pixel `m_remote_message_dismissed` sent
	• Relaunch the app 
	• Ensure a new message appears
	• Click on "Get tips"
	• Ensure it performs a search of "covid 19"
	• Ensure `m_remote_message_primary_action_clicked` pixel sent
	• Open a new tab
	• Ensure remote message is gone
	
3. Message with favorites
	• Add 2 sites as favorites. Open new tab and ensure favorites are visible
	• Relaunch the app
	• Ensure new message "app update" appears
	• Ensure remote message and favorites are visible
	• Click on "Later"
	• Ensure `m_remote_message_secondary_action_clicked` pixel sent
	• Ensure remote message is gone

4. Run unit tests and ensure they are green

image image image image

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPad

**OS Testing**:

* [ ] iOS 14
* [ ] iOS 15

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**

### CODE REVIEW

File: DuckDuckGo/DaxDialogs.swift [hunk 1]
```diff
@@ -185,6 +185,13 @@ class DaxDialogs {
         return nonDDGBrowsingMessageSeen && !fireButtonBrowsingMessageSeenOrExpired && isEnabled
     }
 
+    func isStillOnboarding() -> Bool {
+        if settings.homeScreenMessagesSeen < 2 && !settings.isDismissed {
```
Comment (THISISDINOSAUR) [inline DuckDuckGo/DaxDialogs.swift]: I don't love hard coding this 2, it seems like an easy thing to overlook if we ever change anything, is there another way we can do this check?

File: DuckDuckGo/RemoteMessaging.swift [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            return bookmarksCachingSearch.favoritesCount
+        }
+        let isWidgetInstalled = await AppDependencyProvider.shared.appSettings.isWidgetInstalled()
+
+        try await Self.fetchAndProcess(bookmarksCount: bookmarksCount,
+                                       favoritesCount: favoritesCount,
+                                       isWidgetInstalled: isWidgetInstalled)
+    }
+
+    static func fetchAndProcess(bookmarksCount: Int,
```
Comment (THISISDINOSAUR) [inline DuckDuckGo/RemoteMessaging.swift]: Is this used anywhere else? I'd be tempted to make it private

**Review by THISISDINOSAUR — APPROVED** [review]
LGTM! I do have some questions that related to both PRs, but I'll take them to the other one since I have some other questions there too

File: DuckDuckGo/DaxDialogs.swift (hunk 1 above)
Comment (amddg44) [reply DuckDuckGo/DaxDialogs.swift]: Thanks for catching, I meant to replace this. I've re-worked the logic here to something that should be much safer long term + have added to the unit tests for edge cases

File: DuckDuckGo/RemoteMessagingStore.swift [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+        os_log("Remote messaging config - save processed version: %d", log: .remoteMessaging, type: .debug, processorResult.version)
+        saveRemoteMessagingConfig(withVersion: processorResult.version)
+
+        if let remoteMessage = processorResult.message {
+            deleteScheduledRemoteMessages()
+            save(remoteMessage: remoteMessage)
+
+            DispatchQueue.main.async {
+                self.notificationCenter.post(name: RemoteMessaging.Notifications.remoteMessagesDidChange, object: nil)
+            }
```
Comment (bwaresiak) [inline DuckDuckGo/RemoteMessagingStore.swift:59]: What about else here? Shouldn't we remove scheduled messages?

File: DuckDuckGo/RemoteMessagingStore.swift [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+//  limitations under the License.
+//
+
+import Foundation
+import CoreData
+import os.log
+import Core
+import BrowserServicesKit
+
+final class RemoteMessagingStore: RemoteMessagingStoring {
```
Comment (bwaresiak) [inline DuckDuckGo/RemoteMessagingStore.swift:26]: Is this class intended to be thread safe or rather bound to the main thread?

File: DuckDuckGo/RemoteMessagingStore.swift [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+    func fetchRemoteMessagingConfig() -> RemoteMessagingConfig? {
+        var config: RemoteMessagingConfig?
+        context.performAndWait {
+            let fetchRequest: NSFetchRequest<RemoteMessagingConfigManagedObject>
+                    = RemoteMessagingConfigManagedObject.fetchRequest()
+            fetchRequest.sortDescriptors = [NSSortDescriptor(key: "version", ascending: false)]
+            guard let results = try? context.fetch(fetchRequest) else {
+                return
+            }
+            if let remoteMessagingConfigManagedObject = results.first {
```
Comment (bwaresiak) [inline DuckDuckGo/RemoteMessagingStore.swift:81]: For every fetch request where you need only one element, please set fetch limit to 1.

File: DuckDuckGo/RemoteMessagingStore.swift [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+                remoteMessagingConfigManagedObject.evaluationTimestamp = Date()
+                remoteMessagingConfigManagedObject.invalidate = false
+            }
+
+            do {
+                try context.save()
+            } catch {
+                os_log("Failed to save remote messaging config: %@",
+                       log: OSLog.remoteMessaging,
+                       type: .error, error.localizedDescription)
```
Comment (bwaresiak) [inline DuckDuckGo/RemoteMessagingStore.swift:122]: Shall we have some debug pixels for this (and other) DB errors? Especially considering some updates that may happen in the background.

File: DuckDuckGo/RemoteMessagingStore.swift [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+            }
+        }
+    }
+
+    private func invalidateRemoteMessagingConfigs() {
+        context.performAndWait {
+            let fetchRequest: NSFetchRequest<RemoteMessagingConfigManagedObject> = RemoteMessagingConfigManagedObject.fetchRequest()
+            guard let results = try? context.fetch(fetchRequest) else { return }
+
+            results.forEach { $0.invalidate = true }
```
Comment (bwaresiak) [inline DuckDuckGo/RemoteMessagingStore.swift:134]: As we iterate over the full fetched collection, it is a good practice to set return objects as faults to false (here and in other similar fetches in this file).

**Review by bwaresiak — CHANGES_REQUESTED** [review]
Hey, sorry I'm late - great work overall, I've added few comments inline.

File: DuckDuckGo/RemoteMessagingStore.swift (hunk 3 above)
Comment (amddg44) [reply DuckDuckGo/RemoteMessagingStore.swift:59]: 👍 I've made an update here - thanks for catching

File: DuckDuckGo/RemoteMessagingStore.swift (hunk 4 above)
Comment (amddg44) [reply DuckDuckGo/RemoteMessagingStore.swift:26]: 👍 thread safe

File: DuckDuckGo/RemoteMessagingStore.swift (hunk 5 above)
Comment (amddg44) [reply DuckDuckGo/RemoteMessagingStore.swift:81]: no problem, that's done

File: DuckDuckGo/RemoteMessagingStore.swift (hunk 7 above)
Comment (amddg44) [reply DuckDuckGo/RemoteMessagingStore.swift:134]: 👍 that's done thanks

File: DuckDuckGo/RemoteMessagingStore.swift (hunk 6 above)
Comment (amddg44) [reply DuckDuckGo/RemoteMessagingStore.swift:122]: great suggestion @bwaresiak. @cmonfortep I've put in debug error pixels as was suggested here; ping me if you'd like to do the same and we can be consistent across platforms for error reporting

File: DuckDuckGo/RemoteMessagingStore.swift [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+    init() {
+        notificationCenter = .default
+    }
+
+    init(context: NSManagedObjectContext, notificationCenter: NotificationCenter = .default) {
+        self.notificationCenter = notificationCenter
+        self.context = context
+    }
+
+    private lazy var context = Database.shared.makeContext(concurrencyType: .privateQueueConcurrencyType, name: Constants.privateContextName)
```
Comment (bwaresiak) [inline DuckDuckGo/RemoteMessagingStore.swift]: Lazy variables are not thread safe, as their initialization can crash when accessed concurrently.

File: DuckDuckGo/RemoteMessagingStore.swift [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+                dismissed = true
+            }
+        }
+        return dismissed
+    }
+
+    func dismissRemoteMessage(withId id: String) {
+        updateRemoteMessage(withId: id, toStatus: .dismissed)
+        invalidateRemoteMessagingConfigs()
+    }
```
Comment (bwaresiak) [inline DuckDuckGo/RemoteMessagingStore.swift:237]: Since this class supports concurrency, shouldn't these kind of multi-step functions be done in transactional way (to ensure no change of state will "sneak" between these calls)?

**Review by bwaresiak — COMMENTED** [review]
Thanks for clarification & tweaks, I've added two more comments to follow up on this class concurrency.

File: DuckDuckGo/RemoteMessagingStore.swift (hunk 8 above)
Comment (amddg44) [reply DuckDuckGo/RemoteMessagingStore.swift]: That's done

File: DuckDuckGo/RemoteMessagingStore.swift (hunk 9 above)
Comment (amddg44) [reply DuckDuckGo/RemoteMessagingStore.swift:237]: That's done

**Review by bwaresiak — APPROVED** [review]
Thanks @amddg44 !
